Add users colors to the theme

hbcarlos opened the pull request:
Follow-up of #11443

## References
#11641

## Code changes
Adds new variables to the theme for the different colors and change the list of colors in the user package to a list of variables.

## User-facing changes

## Backwards-incompatible changes

jtpio reviewed (approved):
Thanks!

jtpio commented:
cc @ellisonbg @krassowski since they are mentioned in the related issue: https://github.com/jupyterlab/jupyterlab/issues/11641#issue-1074802240

ellisonbg commented:
Thanks for this follow up PR. The code looks fine, my only questions are around how many colors there are (why exactly 8?) and what colors we are using from a design and accessibility perspective. To help reviewers review these questions, can you post a screenshot of the color sequence, or provide a link to the colors? Where did this color sequence come from originally?

ellisonbg commented:
I forgot I have some color palette visualization code in a notebook. Here is the proposed color sequence, along with WCAG 2.x contrast for white and black text:

<img width="331" alt="Screen Shot 2021-12-16 at 5 27 29 PM" src="https://user-images.githubusercontent.com/27600/146473708-bc2f7b11-6fc0-495c-bd31-61bc2e096822.png">

While these all have good contrast as a background for black text, the contrast between some pairs is very small. In particular (1,6,8) and (2,7) are very similar and will be difficult to distinguish.

ellisonbg commented:
OK, I have written some code that can generate categorical color palettes using the following approach:

* Work in HCL space
* Set the Luminance (L) to a fixed value that all the colors meet WCAG contrast goals and have the same visual weight in the application.
* Set the Chroma (C) to a fixed value based on visual design.
* Use a set of equally spaces Hue (H) values, with `N` points.
* Optionally rotate all `N` colors in Hue space based on visual design.
* Allow the color library I am using (colorspace) to apply corrections to bring out of gamut colors back into sRGB space.

Here are few examples:

N=7, C=80, L=80 => good contrast with black text:

<img width="333" alt="Screen Shot 2021-12-16 at 6 04 11 PM" src="https://user-images.githubusercontent.com/27600/146476774-1947d481-5fbf-4f60-959c-95181af5cb1d.png">

N=7, C=100, L=30 => good contrast with white text:

<img width="333" alt="Screen Shot 2021-12-16 at 6 09 05 PM" src="https://user-images.githubusercontent.com/27600/146477223-ad1e749f-33b8-48fc-9770-9556fab2082f.png">

After playing around with this for a while, my sense is that it is easier to generate a palette that works well with black text, as the lower luminance value begin to pull colors towards being "muddy" and dark.

In terms of the overall number of colors, I can increase `N`, but there is a tradeoff as adjacent colors become more similar.

ellisonbg commented:
I can also share the code with anyone who wants to play with this further.

hbcarlos commented:
Hi @ellisonbg, thank you for creating t a new palette for both themes!
I'm adding them to both themes.

> how many colors there are (why exactly 8?) and what colors we are using from a design and accessibility perspective.
> Where did this color sequence come from originally?

There are eight colors because is the colors we decided to use in https://github.com/jupyterlab/jupyterlab/pull/10411, and @krassowski tested them in https://github.com/jupyterlab/jupyterlab/pull/10411#discussion_r651960093

krassowski commented:
@ellisonbg I appreciate the effort you put into generating a palette of colours with good contrast with text. I am not convinced by the current result - the proposed colours are too similar to each other for me (I have a mild colourblindness though).

Could we go a step back, and define what we try to achieve and what the design constraints are? For one I will advocate to add a strong constraint on how distinct the colours are (I see that you already noted this problem in https://github.com/jupyterlab/jupyterlab/pull/11703#issuecomment-996355204) and what contrast we want to achieve. I don't think that we need so high contrast (>>5) nor that we need to check contrast against black text (for the light theme) at all - because such combination is not used in any application of colours to RTC that I encountered in the wild (see below).

I see the following design constraints for Light Theme and following consequences:
1. the cursor has to be distinct against the background which is white/light grey → cursor has to be dark
2. the name on the cursor needs to contrast against such dark cursor → the text needs to be white
3. we want to associate the colour with user in other elements of the UI (e.g. collaborator icons) which means that we need to use:
   - a) the same colour as background and white text in the collaborator icon, or
   - b) the same colour as background and black text with white shadow/outline, or
   - c) a variation of the same colour but with lighter luminosity as background and black text (downside: harder to connect the cursor to collaborator icon as the colours differ), or
   - d) do not use colour as background but use it as an accent on the user icon e.g.:
       - add a border
       - add an underline
       - add a small secondary icon
4. users want to refer to cursors of other people with words → we need to maximise _color saliency_ (=choose colours which have universally understood names so that users can refer to them easily; [(Heer and Stone, 2012)](https://research.tableau.com/sites/default/files/2012-ColorNameModels-CHI.pdf) paper gives an excellent overview of this topic and provides a [website](http://vis.stanford.edu/color-names/) for testing _color saliency_ and _color name distance_)
5. users want to distinguish cursors of different people → we want to:
    - maximise the number of colours so that they do not repeat,
    - maximise distance between colors in a color space of choice so they are visually distinct, and
    - maximise the _colour name distance_ (to satisfy the point four, above)
6. ideally, the colour palette would not be associated with _negative_, or _disturbing_ affective interpretations → avoid colour palettes associated with those interpretations as in [(Bartram et al, 2017)](https://research.tableau.com/sites/default/files/Affective%20Color%20CHI%202017.pdf)

I like the old colours but I guess that other colours from Material palette that we already use would work too; it seems that all RTC applications use a similar palette without going as dark, nor as pastel as the colours proposed in https://github.com/jupyterlab/jupyterlab/pull/11703#issuecomment-996378177; they also all use a variation of (1), (2) and (3) rules (and I believe Google Docs also meets rules (4), (5) and (6)):

<details>

- Google Docs uses (1) and (2) and (3a) for anonymous users and (3d) for users with a profile photo:
    > ![Screenshot from 2021-12-28 13-49-51](https://user-images.githubusercontent.com/5832902/147573091-16dbd89b-fe67-4e3a-8c5f-e0cf60900307.png)
    > ![Screenshot from 2021-12-28 13-49-45](https://user-images.githubusercontent.com/5832902/147573095-856e1f51-b4c8-4267-adcb-c231304be1f7.png)
- Atlassian uses (1) and (2) and (3d) with (3a):
    > ![image](https://user-images.githubusercontent.com/5832902/147572326-41adb278-d2c4-4d2b-982f-2829008911c7.png)
- Deepnote uses (1) and (2) an (3d):
    > ![Screenshot from 2021-12-28 13-55-38](https://user-images.githubusercontent.com/5832902/147574105-70135149-4cb4-4dad-9d91-19fd039c4421.png)
    > ![Screenshot from 2021-12-28 13-54-59](https://user-images.githubusercontent.com/5832902/147574107-bebcebdb-371b-41fc-a066-ea5e15f7ac33.png)
- Microsoft Word 365 uses (1) and (2) and (3a):
    > ![Screenshot from 2021-12-28 14-08-09](https://user-images.githubusercontent.com/5832902/147574905-715bff1c-6e9a-4353-b433-b96ee9ba8772.png)
    > ![Screenshot from 2021-12-28 14-14-39](https://user-images.githubusercontent.com/5832902/147575125-61ae5b14-3bf8-4f00-888d-bf7bdbaad249.png)
- New versions of MS Word will use (3c) but with a dark color not black, and (3d) (based on [this news](https://www.computerworld.com/article/3635689/what-s-new-in-office-2021.html)):
    ![Screenshot from 2021-12-28 14-02-35](https://user-images.githubusercontent.com/5832902/147574286-a18d934f-fd02-468d-8205-3c24e6d7fcdf.png)

</details>

Importantly, none of the above uses black text with the cursor colour, not even with shadow or outline as suggested in (3b), with the new version of MS world being the only one using (3c).

I propose the following cursor colour selection rules for Light Theme.
1) contrast with the inactive cell editor background (`--jp-layout-color3`, `#bdbdbd` in light theme) should be 4.5:1 or higher
2) contrast with white username text should be 4.5:1 or higher
3) for a given number of colours N maximise the distance between colours within the above constraints in a given colour space
4) maximise _color saliency_ and _color name distance_ using [this analyzer](http://vis.stanford.edu/color-names/analyzer/):
   - calculate saliency and distance for initial results from (4)
   - if the result is not satisfactory increment N and repeat step (4),
   - discard the colour(s) with lowest saliency/contributing the most to reduction in name distance, so that we are back at N colours,
   - check saliency and name distance again,
   - repeat until we have a satisfactory result
5) check the resulting colour palettes affective interpretations; I do not know of a tool for that but we can compare against [(Bartram et al, 2017)](https://research.tableau.com/sites/default/files/Affective%20Color%20CHI%202017.pdf)

For Dark Theme the process would be analogous, but starting from different value for the inactive cell background which is `#616161` by default.

ellisonbg commented:
Thanks for the detailed and thoughtful response, I am just plugging back
into work and will read through your feedback in detail.

On Tue, Dec 28, 2021 at 7:43 AM Michał Krassowski ***@***.***>
wrote:

> @ellisonbg <https://github.com/ellisonbg> I appreciate the effort you put
> into generating a palette of colours with good contrast with text. I am not
> convinced by the current result - the proposed colours are too similar to
> each other for me (I have a mild colourblindness though).
>
> Could we go a step back, and define what we try to achieve and what the
> design constraints are? For one I will advocate to add a strong constraint
> on how distinct the colours are (I see that you already noted this problem
> in #11703 (comment)
> <https://github.com/jupyterlab/jupyterlab/pull/11703#issuecomment-996355204>)
> and what contrast we want to achieve. I don't think that we need so high
> contrast (>>5) nor that we need to check contrast against black text (for
> the light theme) at all - because such combination is not used in any
> application of colours to RTC that I encountered in the wild (see below).
>
> I see the following design constraints for Light Theme and following
> consequences:
>
>    1. the cursor has to be distinct against the background which is
>    white/light grey → cursor has to be dark
>    2. the name on the cursor needs to contrast against such dark cursor →
>    the text needs to be white
>    3. we want to associate the colour with user in other elements of the
>    UI (e.g. collaborator icons) which means that we need to use:
>       - a) the same colour as background and white text in the
>       collaborator icon, or
>       - b) the same colour as background and black text with white
>       shadow/outline, or
>       - c) a variation of the same colour but with lighter luminosity as
>       background and black text (downside: harder to connect the cursor to
>       collaborator icon as the colours differ), or
>       - d) do not use colour as background but use it as an accent on the
>       user icon e.g.:
>          - add a border
>          - add an underline
>          - add a small secondary icon
>       4. users want to refer to cursors of other people with words → we
>    need to maximise *color saliency* (=choose colours which have
>    universally understood names so that users can refer to them easily; (Heer
>    and Stone, 2012)
>    <https://research.tableau.com/sites/default/files/2012-ColorNameModels-CHI.pdf>
>    paper gives an excellent overview of this topic and provides a website
>    <http://vis.stanford.edu/color-names/> for testing *color saliency*
>    and *color name distance*)
>    5. users want to distinguish cursors of different people → we want to:
>       - maximise the number of colours so that they do not repeat,
>       - maximise distance between colors in a color space of choice so
>       they are visually distinct, and
>       - maximise the *colour name distance* (to satisfy the point four,
>       above)
>    6. ideally, the colour palette would not be associated with *negative*,
>    or *disturbing* affective interpretations → avoid colour palettes
>    associated with those interpretations as in (Bartram et al, 2017)
>    <https://research.tableau.com/sites/default/files/Affective%20Color%20CHI%202017.pdf>
>
> I like the old colours but I guess that other colours from Material
> palette that we already use would work too; it seems that all RTC
> applications use a similar palette without going as dark, nor as pastel as
> the colours proposed in #11703 (comment)
> <https://github.com/jupyterlab/jupyterlab/pull/11703#issuecomment-996378177>;
> they also all use a variation of (1), (2) and (3) rules (and I believe
> Google Docs also meets rules (4), (5) and (6)):
>
>    - Google Docs uses (1) and (2) and (3a) for anonymous users and (3d)
>    for users with a profile photo:
>
>    [image: Screenshot from 2021-12-28 13-49-51]
>    <https://user-images.githubusercontent.com/5832902/147573091-16dbd89b-fe67-4e3a-8c5f-e0cf60900307.png>
>    [image: Screenshot from 2021-12-28 13-49-45]
>    <https://user-images.githubusercontent.com/5832902/147573095-856e1f51-b4c8-4267-adcb-c231304be1f7.png>
>
>    - Atlassian uses (1) and (2) and (3d) with (3a):
>
>    [image: image]
>    <https://user-images.githubusercontent.com/5832902/147572326-41adb278-d2c4-4d2b-982f-2829008911c7.png>
>
>    - Deepnote uses (1) and (2) an (3d):
>
>    [image: Screenshot from 2021-12-28 13-55-38]
>    <https://user-images.githubusercontent.com/5832902/147574105-70135149-4cb4-4dad-9d91-19fd039c4421.png>
>    [image: Screenshot from 2021-12-28 13-54-59]
>    <https://user-images.githubusercontent.com/5832902/147574107-bebcebdb-371b-41fc-a066-ea5e15f7ac33.png>
>
>    - Microsoft Word 365 uses (1) and (2) and (3a):
>
>    [image: Screenshot from 2021-12-28 14-08-09]
>    <https://user-images.githubusercontent.com/5832902/147574905-715bff1c-6e9a-4353-b433-b96ee9ba8772.png>
>    [image: Screenshot from 2021-12-28 14-14-39]
>    <https://user-images.githubusercontent.com/5832902/147575125-61ae5b14-3bf8-4f00-888d-bf7bdbaad249.png>
>
>    - New versions of MS Word will use (3c) but with a dark color not
>    black, and (3d) (based on this news
>    <https://www.computerworld.com/article/3635689/what-s-new-in-office-2021.html>
>    ):
>    [image: Screenshot from 2021-12-28 14-02-35]
>    <https://user-images.githubusercontent.com/5832902/147574286-a18d934f-fd02-468d-8205-3c24e6d7fcdf.png>
>
> Importantly, none of the above uses black text with the cursor colour, not
> even with shadow or outline as suggested in (3b), with the new version of
> MS world being the only one using (3c).
>
> I propose the following cursor colour selection rules for Light Theme.
>
>    1. contrast with the inactive cell editor background (
>    --jp-layout-color3, #bdbdbd in light theme) should be 4.5:1 or higher
>    2. contrast with white username text should be 4.5:1 or higher
>    3. for a given number of colours N maximise the distance between
>    colours within the above constraints in a given colour space
>    4. maximise *color saliency* and *color name distance* using this
>    analyzer <http://vis.stanford.edu/color-names/analyzer/>:
>       - calculate saliency and distance for initial results from (4)
>       - if the result is not satisfactory increment N and repeat step (4),
>       - discard the colour(s) with lowest saliency/contributing the most
>       to reduction in name distance, so that we are back at N colours,
>       - check saliency and name distance again,
>       - repeat until we have a satisfactory result
>    5. check the resulting colour palettes affective interpretations; I do
>    not know of a tool for that but we can compare against (Bartram et al,
>    2017)
>    <https://research.tableau.com/sites/default/files/Affective%20Color%20CHI%202017.pdf>
>
> For Dark Theme the process would be analogous, but starting from different
> value for the inactive cell background which is #616161 by default.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jupyterlab/jupyterlab/pull/11703#issuecomment-1002165166>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAGXUH4CML3E5JOEUMJD23UTHLJDANCNFSM5KF6MGQA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Brian E. Granger

Senior Principal Technologist, AWS AI/ML ***@***.***)
On Leave - Professor of Physics and Data Science, Cal Poly
@ellisonbg on GitHub

ellisonbg commented:
Thanks for the ping!

I think it is helpful to think about design goals/constraints. Looking through the ones above:

* I think that 1, 2, 3, 5 make sense.
* Given that our usage of these colors would be accompanied by the users actual name, I am not convinced that color saliency is a strong requirement in this case. Put more simply, why would a user say "the red prompt" when they can read the person's name and say "Steve's prompt"? Collaboration is always based on trust, so users generally make an effort to be as personal as possible.
* I don't have a problem with 6, but am not clear on exactly how we build a palette with that in mind while satisfying the other constraints.

In addition to that, I would add one more: that all the colors have roughly the same Luminance value. Each of these colors will be used in the sames contexts, and we would want the visual contrast with background elements and text to be roughly the same, so I think this is required.

The only other feedback I have is that I am not clear on how write code or make human driven decisions that build a palette that satisfies these constraints. I believe the second palette I generated (or one like it with perhaps slightly lighter colors - we don't need quite that high constrast against white text) mostly does the job (I guess I am saying I don't know how to do better). Here is an example of one such palette:

C=90, L=44, N=7:

![Screen Shot 2022-01-14 at 12 49 23 PM](https://user-images.githubusercontent.com/27600/149583241-56fb5b4c-ae16-430f-adcc-af95418595d5.png)

I should note that it is relatively easy for us to iterate on colors over time, however, once we fix the **number** of colors, it is more difficult to change (especially if we want to reduce the number) without breaking the API. Given that, I would propose we finalize the number of colors and pick a reasonable candidate for the colors and move forward.

hbcarlos commented:
I rebased and changed the dark theme palette for the one posted in the comment https://github.com/jupyterlab/jupyterlab/pull/11703#issuecomment-1013465331.

This is ready for review.

bollwyvl commented on packages/theme-dark-extension/style/variables.css:
```diff
@@ -369,6 +369,20 @@ all of MD as it is not optimized for dense, information rich UIs.
+    and the icon of the user.
+  */
+
+  --jp-collaborator-1: #ad4a00;
```
perhaps `--jp-collaborator-color1` for parity with other naming schemes? presumably we could add some other distinguisher (e.g. border) that would give us 7 x 7 unique combinations or something.

hbcarlos replied:
That's a great idea. Thanks, @bollwyvl.
I'll apply the changes!

fcollonval reviewed (approved):
Thanks @hbcarlos

fcollonval commented:
Sorry for the delay @hbcarlos - merging...
